File: parser/src/migration.ts
import * as crypto from "crypto";
import * as fs from "fs";
import * as path from "path";

export enum MigrationStatus {
  PENDING = "pending",
  RUNNING = "running",
  COMPLETED = "completed",
  FAILED = "failed",
  ROLLED_BACK = "rolled_back",
}

export enum UpgradeStrategy {
  BLUE_GREEN = "blue_green",
  ROLLING = "rolling",
  CANARY = "canary",
  BIG_BANG = "big_bang",
}

export interface UpgradeManifest {
  versionFrom: string;
  versionTo: string;
  strategy: UpgradeStrategy;
  steps: Record<string, unknown>[];
  rollbackSteps: Record<string, unknown>[];
  metadata: Record<string, unknown>;
  _id?: string;
}

export interface MigrationRecord {
  manifestId: string;
  status: MigrationStatus;
  currentStep: number;
  logs: string[];
  startedAt?: string;
  completedAt?: string;
  error?: string;
}

export class MigrationEngine {
  private checkpointDir: string;
  private activeMigrations: Map<string, MigrationRecord> = new Map();
  private versionLocks: Map<string, string> = new Map();

  constructor(checkpointDir: string = ".alp/migrations") {
    this.checkpointDir = checkpointDir;
    if (!fs.existsSync(checkpointDir)) {
      fs.mkdirSync(checkpointDir, { recursive: true });
    }
  }

  registerMigration(manifest: UpgradeManifest): string {
    const id = manifest._id ?? this.computeManifestId(manifest);
    const record: MigrationRecord = {
      manifestId: id,
      status: MigrationStatus.PENDING,
      logs: [],
      currentStep: 0,
    };
    this.activeMigrations.set(id, record);
    this.saveCheckpoint(record);
    return id;
  }

  execute(manifest: UpgradeManifest, dryRun: boolean = false): MigrationRecord {
    const id = manifest._id ?? this.computeManifestId(manifest);
    let record = this.activeMigrations.get(id);
    if (!record) {
      record = {
        manifestId: id,
        status: MigrationStatus.RUNNING,
        logs: [],
        currentStep: 0,
      };
      this.activeMigrations.set(id, record);
    }

    record.status = MigrationStatus.RUNNING;
    record.startedAt = this.now();
    record.logs.push(
      `Starting migration ${manifest.versionFrom} -> ${manifest.versionTo}`
    );

    if (dryRun) {
      record.logs.push("DRY RUN: validating steps only");
      manifest.steps.forEach((step, i) => {
        record.logs.push(`Step ${i + 1}: ${step["name"] ?? "unnamed"} [validated]`);
      });
      record.status = MigrationStatus.COMPLETED;
      record.completedAt = this.now();
      this.saveCheckpoint(record);
      return record;
    }

    try {
      this.versionLocks.set(manifest.versionFrom, manifest.versionTo);
      manifest.steps.forEach((step, i) => {
        record.currentStep = i;
        const stepName = step["name"] ?? `step_${i}`;
        record.logs.push(`Executing step ${i + 1}: ${stepName}`);
        this.executeStep(step);
      });
      record.status = MigrationStatus.COMPLETED;
      record.completedAt = this.now();
      record.logs.push(`Migration completed: ${manifest.versionTo}`);
    } catch (e) {
      record.status = MigrationStatus.FAILED;
      record.error = String(e);
      record.logs.push(`Migration failed: ${e}`);
      this.rollback(manifest, record);
    }

    this.saveCheckpoint(record);
    return record;
  }

  rollback(manifest: UpgradeManifest, record: MigrationRecord): MigrationRecord {
    return this.rollbackInternal(manifest, record);
  }

  private rollbackInternal(
    manifest: UpgradeManifest,
    record: MigrationRecord
  ): MigrationRecord {
    record.status = MigrationStatus.ROLLED_BACK;
    record.logs.push("Initiating rollback...");
    manifest.rollbackSteps.forEach((step, i) => {
      const stepName = step["name"] ?? `rollback_step_${i}`;
      record.logs.push(`Rollback step ${i + 1}: ${stepName}`);
      this.executeStep(step);
    });
    record.logs.push("Rollback completed");
    if (!record.completedAt) {
      record.completedAt = this.now();
    }
    this.saveCheckpoint(record);
    return record;
  }

  private executeStep(step: Record<string, unknown>): void {
    const stepType = step["type"] ?? "noop";
    if (stepType === "noop") {
      return;
    } else if (stepType === "callback" && typeof step["fn"] === "function") {
      (step["fn"] as () => void)();
    } else if (stepType === "shell" && typeof step["command"] === "string") {
      require("child_process").execSync(step["command"], {
        stdio: "ignore",
      });
    }
  }

  getStatus(migrationId: string): MigrationRecord | undefined {
    return this.activeMigrations.get(migrationId);
  }

  listMigrations(): MigrationRecord[] {
    return Array.from(this.activeMigrations.values());
  }

  getVersionLock(version: string): string | undefined {
    return this.versionLocks.get(version);
  }

  saveCheckpoint(record: MigrationRecord): void {
    const filePath = path.join(this.checkpointDir, `${record.manifestId}.json`);
    const data = {
      manifestId: record.manifestId,
      status: record.status,
      currentStep: record.currentStep,
      logs: record.logs,
      startedAt: record.startedAt,
      completedAt: record.completedAt,
      error: record.error,
    };
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), "utf-8");
  }

  loadCheckpoint(migrationId: string): MigrationRecord | null {
    const filePath = path.join(this.checkpointDir, `${migrationId}.json`);
    if (!fs.existsSync(filePath)) {
      return null;
    }
    const data = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    return {
      manifestId: data.manifestId,
      status: data.status,
      currentStep: data.currentStep ?? 0,
      logs: data.logs ?? [],
      startedAt: data.startedAt,
      completedAt: data.completedAt,
      error: data.error,
    };
  }

  private computeManifestId(manifest: UpgradeManifest): string {
    const content = `${manifest.versionFrom}:${manifest.versionTo}:${manifest.strategy}`;
    return crypto.createHash("sha256").update(content).digest("hex").slice(0, 16);
  }

  private now(): string {
    return new Date().toISOString();
  }
}

File: sdk/python/alp_sdk/migration.py
from __future__ import annotations

import hashlib
import json
import os
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Callable
from enum import Enum


class MigrationStatus(Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    ROLLED_BACK = "rolled_back"


class UpgradeStrategy(Enum):
    BLUE_GREEN = "blue_green"
    ROLLING = "rolling"
    CANARY = "canary"
    BIG_BANG = "big_bang"


@dataclass
class UpgradeManifest:
    version_from: str
    version_to: str
    strategy: UpgradeStrategy
    steps: List[Dict[str, Any]] = field(default_factory=list)
    rollback_steps: List[Dict[str, Any]] = field(default_factory=list)
    metadata: Dict[str, Any] = field(default_factory=dict)
    _id: Optional[str] = None

    def __post_init__(self):
        if self._id is None:
            content = f"{self.version_from}:{self.version_to}:{self.strategy.value}"
            self._id = hashlib.sha256(content.encode()).hexdigest()[:16]


@dataclass
class MigrationRecord:
    manifest_id: str
    status: MigrationStatus
    current_step: int = 0
    logs: List[str] = field(default_factory=list)
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    error: Optional[str] = None


class MigrationEngine:
    def __init__(self, checkpoint_dir: str = ".alp/migrations"):
        self.checkpoint_dir = checkpoint_dir
        self._active_migrations: Dict[str, MigrationRecord] = {}
        self._version_locks: Dict[str, str] = {}
        os.makedirs(checkpoint_dir, exist_ok=True)

    def register_migration(self, manifest: UpgradeManifest) -> str:
        record = MigrationRecord(
            manifest_id=manifest._id,
            status=MigrationStatus.PENDING,
        )
        self._active_migrations[manifest._id] = record
        self._save_checkpoint(record)
        return manifest._id

    def execute(self, manifest: UpgradeManifest, dry_run: bool = False) -> MigrationRecord:
        migration_id = manifest._id
        record = self._active_migrations.get(migration_id)
        if record is None:
            record = MigrationRecord(
                manifest_id=migration_id,
                status=MigrationStatus.RUNNING,
            )
            self._active_migrations[migration_id] = record

        record.status = MigrationStatus.RUNNING
        record.started_at = self._now()
        record.logs.append(f"Starting migration {manifest.version_from} -> {manifest.version_to}")

        if dry_run:
            record.logs.append("DRY RUN: validating steps only")
            for i, step in enumerate(manifest.steps):
                record.logs.append(f"Step {i + 1}: {step.get('name', 'unnamed')} [validated]")
            record.status = MigrationStatus.COMPLETED
            record.completed_at = self._now()
            self._save_checkpoint(record)
            return record

        try:
            self._version_locks[manifest.version_from] = manifest.version_to
            for i, step in enumerate(manifest.steps):
                record.current_step = i
                step_name = step.get("name", f"step_{i}")
                record.logs.append(f"Executing step {i + 1}: {step_name}")
                self._execute_step(step)
            record.status = MigrationStatus.COMPLETED
            record.completed_at = self._now()
            record.logs.append(f"Migration completed: {manifest.version_to}")
        except Exception as e:
            record.status = MigrationStatus.FAILED
            record.error = str(e)
            record.logs.append(f"Migration failed: {e}")
            self._rollback(manifest, record)

        self._save_checkpoint(record)
        return record

    def rollback(self, manifest_id: str) -> MigrationRecord:
        manifest = self._find_manifest(manifest_id)
        if manifest is None:
            raise ValueError(f"Migration {manifest_id} not found")
        record = self._active_migrations.get(manifest_id)
        if record is None:
            raise ValueError(f"No active migration found for {manifest_id}")
        return self._rollback(manifest, record)

    def _rollback(self, manifest: UpgradeManifest, record: MigrationRecord) -> MigrationRecord:
        record.status = MigrationStatus.ROLLED_BACK
        record.logs.append("Initiating rollback...")
        for i, step in enumerate(reversed(manifest.rollback_steps)):
            step_name = step.get("name", f"rollback_step_{i}")
            record.logs.append(f"Rollback step {i + 1}: {step_name}")
            self._execute_step(step)
        record.logs.append("Rollback completed")
        if record.completed_at is None:
            record.completed_at = self._now()
        self._save_checkpoint(record)
        return record

    def _execute_step(self, step: Dict[str, Any]) -> None:
        step_type = step.get("type", "noop")
        if step_type == "noop":
            return
        elif step_type == "callback" and "fn" in step:
            step["fn"]()
        elif step_type == "shell" and "command" in step:
            os.system(step["command"])
        elif step_type == "python" and "code" in step:
            exec(step["code"])
        else:
            pass

    def get_status(self, migration_id: str) -> Optional[MigrationRecord]:
        return self._active_migrations.get(migration_id)

    def list_migrations(self) -> List[MigrationRecord]:
        return list(self._active_migrations.values())

    def get_version_lock(self, version: str) -> Optional[str]:
        return self._version_locks.get(version)

    def _save_checkpoint(self, record: MigrationRecord) -> None:
        path = os.path.join(self.checkpoint_dir, f"{record.manifest_id}.json")
        data = {
            "manifest_id": record.manifest_id,
            "status": record.status.value,
            "current_step": record.current_step,
            "logs": record.logs,
            "started_at": record.started_at,
            "completed_at": record.completed_at,
            "error": record.error,
        }
        with open(path, "w") as f:
            json.dump(data, f, indent=2)

    def _load_checkpoint(self, migration_id: str) -> Optional[MigrationRecord]:
        path = os.path.join(self.checkpoint_dir, f"{migration_id}.json")
        if not os.path.exists(path):
            return None
        with open(path, "r") as f:
            data = json.load(f)
        return MigrationRecord(
            manifest_id=data["manifest_id"],
            status=MigrationStatus(data["status"]),
            current_step=data.get("current_step", 0),
            logs=data.get("logs", []),
            started_at=data.get("started_at"),
            completed_at=data.get("completed_at"),
            error=data.get("error"),
        )

    def _find_manifest(self, migration_id: str) -> Optional[UpgradeManifest]:
        for record in self._active_migrations.values():
            if record.manifest_id == migration_id:
                return UpgradeManifest(
                    version_from="unknown",
                    version_to="unknown",
                    strategy=UpgradeStrategy.BLUE_GREEN,
                    _id=migration_id,
                )
        return None

    @staticmethod
    def _now() -> str:
        import datetime
        return datetime.datetime.utcnow().isoformat() + "Z"
